graphdatascience/aura_api.py

```diff
@@ -0,0 +1,218 @@
+            self.access_token = json["access_token"]
+            expires_in: int = json["expires_in"]
+            self.expires_at = int(time.time()) + expires_in
+            self.token_type = json["token_type"]
+
+        def is_expired(self) -> bool:
+            return self.expires_at >= int(time.time())
+
+    def __init__(self, client_id: str, client_secret: str, tenant_id: Optional[str] = None) -> None:
+        self._credentials = (client_id, client_secret)
+        self._token: Optional[AuraApi.AuraAuthToken] = None
+        self._logger = logging.getLogger()
+        self._tenant_id = tenant_id if tenant_id else self._get_tenant_id()
+
+    def create_instance(self, name: str) -> InstanceCreateDetails:
+        # TODO should give more control here
+        data = {
+            "name": name,
+            "memory": "8GB",
+            "version": "5",
+            "region": "europe-west1",
+            # TODO should be figured out from the tenant details in the future
+            "type": "enterprise-ds" if not AuraApi.DEV_ENV else "professional-ds",
+            "tenant_id": self._tenant_id,
+            "cloud_provider": "gcp",
+        }
+
+        response = req.post(
+            f"{AuraApi.BASE_URI}/v1/instances",
+            json=data,
+            headers=self._get_request_headers(),
+        )
+
+        try:
+            response.raise_for_status()
+        except HTTPError as e:
+            print(response.json())
+            raise e
+
+        return InstanceCreateDetails.fromJson(response.json()["data"])
+
+    def delete_instance(self, instance_id: str) -> Optional[InstanceSpecificDetails]:
+        response = req.delete(
+            f"{AuraApi.BASE_URI}/v1/instances/{instance_id}",
+            headers=self._get_request_headers(),
+        )
+
+        if response.status_code == 404:
+            return None
+
+        response.raise_for_status()
+
+        return InstanceSpecificDetails.fromJson(response.json()["data"])
+
+    def list_instances(self) -> List[InstanceDetails]:
+        response = req.get(
+            f"{AuraApi.BASE_URI}/v1/instances",
+            headers=self._get_request_headers(),
+            params={"tenantId": self._tenant_id},
+        )
+
+        response.raise_for_status()
+
+        raw_data = response.json()["data"]
+
+        return [InstanceDetails.fromJson(i) for i in raw_data]
+
+    def list_instance(self, instance_id: str) -> Optional[InstanceSpecificDetails]:
+        response = req.get(
+            f"{AuraApi.BASE_URI}/v1/instances/{instance_id}",
+            headers=self._get_request_headers(),
+        )
+
+        if response.status_code == 404:
+            return None
+
+        response.raise_for_status()
+
+        raw_data = response.json()["data"]
+
+        return InstanceSpecificDetails.fromJson(raw_data)
+
+    def wait_for_instance_running(self, instance_id: str, sleep_time: float = 0.2) -> bool:
+        while sleep_time < AuraApi.MAX_WAIT_TIME:
```

Comment:
The `sleep_time`variable is never updated resulting in a possibly endless waiting loop.

Reply:
nice find! there's no unit test for this method 🤦

Reply:
Added a few tests now